diff --git a/server/mpm/prefork/prefork.c b/server/mpm/prefork/prefork.c
index 6962026..1b54b4e 100644
--- a/server/mpm/prefork/prefork.c
+++ b/server/mpm/prefork/prefork.c
@@ -687,6 +687,11 @@ static void child_main(int child_num_arg)
             die_now = 1;
         }
     }
+    /* This apr_pool_clear call is redundant, should be redundant, but compensates
+     * a flaw in the apr reslist code.  This should be removed once that flaw has
+     * been addressed.
+     */
+    apr_pool_clear(ptrans);
     clean_child_exit(0);
 }
 
