diff --git a/server/scoreboard.c b/server/scoreboard.c
index eb683b3..6347311 100644
--- a/server/scoreboard.c
+++ b/server/scoreboard.c
@@ -505,8 +505,8 @@ void ap_time_process_request(ap_sb_handle_t *sbh, int status)
 
 AP_DECLARE(worker_score *) ap_get_scoreboard_worker(int x, int y)
 {
-    if (((x < 0) || (server_limit < x)) ||
-        ((y < 0) || (thread_limit < y))) {
+    if (((x < 0) || (x >= server_limit)) ||
+        ((y < 0) || (y >= thread_limit))) {
         return(NULL); /* Out of range */
     }
     return &ap_scoreboard_image->servers[x][y];
@@ -529,7 +529,7 @@ AP_DECLARE(void) ap_copy_scoreboard_worker(worker_score *dest,
 
 AP_DECLARE(process_score *) ap_get_scoreboard_process(int x)
 {
-    if ((x < 0) || (server_limit < x)) {
+    if ((x < 0) || (x >= server_limit)) {
         return(NULL); /* Out of range */
     }
     return &ap_scoreboard_image->parent[x];
@@ -542,7 +542,7 @@ AP_DECLARE(global_score *) ap_get_scoreboard_global()
 
 AP_DECLARE(lb_score *) ap_get_scoreboard_lb(int lb_num)
 {
-    if (((lb_num < 0) || (lb_limit < lb_num))) {
+    if (lb_num < 0 || lb_num >= lb_limit) {
         return(NULL); /* Out of range */
     }
     return &ap_scoreboard_image->balancers[lb_num];
